### PR TITLE
Add initial `spack.yaml` and `config`

### LINKED ISSUE #2: Add initial `spack.yaml` and `config/versions.json`
Add an initial `spack.yaml` and `config/versions.json` to allow for initial testing of `ACCESS-ESM1.5`. 

Requires #1 to be merged first.

### PR DESCRIPTION
**Pull request by CodeGat**
In this PR:
* Add initial `spack.yaml`
* Add initial `config` directory with versions of `spack`, `spack-packages` and `spack-config`

**Note**: https://github.com/ACCESS-NRI/ACCESS-ESM1.5/pull/7 must be merged first, so the build-db metadata scraping works correctly. 

Closes #2

### CODE REVIEW

**Comment by CodeGat** [comment]
Gadi is unhappy at the moment, so the deployment has failed. Will rerun when it is better. As an aside...maybe it is worth upserting or deleting old comments...the PRs become a bit large with all these comments...
See failed run: https://github.com/ACCESS-NRI/ACCESS-ESM1.5/actions/runs/9345587853/job/25718740970?pr=5

**Comment by harshula** [comment]
Hi @CodeGat , I think we should require an exact version of netcdf* and hdf5 packages that correspond with the non-Spack build.

**Comment by CodeGat** [comment]
Fair enough @harshula - what would those versions be, or where can I find them?

**Comment by harshula** [comment]
> what would those versions be, or where can I find them?

Hi @penguian , Can you please list here, the versions of netcdf* and hdf5 packages that correspond with the non-Spack build?

**Comment by penguian** [comment]
Versions used by the non-Spack build at https://github.com/penguian/access-esm-build-gadi/blob/master/Makefile : 
* `netcdf/4.7.1` -- from `/apps/Modules/modulefiles` and `/apps/netcdf/4.7.1/lib/pkgconfig/netcdf.pc`, with settings at `/apps/netcdf/4.7.1/lib/libnetcdf.settings`
* `netcdf-fortran` 4.5.1 -- from `/apps/netcdf/4.7.1/lib/Intel/pkgconfig/netcdf-fortran.pc` 
* `hdf5/1.10.5` -- from `/apps/Modules/modulefiles` and /apps/netcdf/4.7.1/lib/libnetcdf.la

**Comment by penguian** [comment]
Note that MOM5 for ACCESS-OM2 uses `netcdf/4.7.4`: https://github.com/mom-ocean/MOM5/pull/333
I don't think that it would hurt to use 4.7.4 for ACCESS-ESM1.5 as well, but we need to check.

**Comment by CodeGat** [comment]
Thanks Paul. I'll add those versions, and try with 4.7.4. By `netcdf`, do you mean the spack package `netcdf-c`?

**Comment by penguian** [comment]
> Thanks Paul. I'll add those versions, and try with 4.7.4. By `netcdf`, do you mean the spack package `netcdf-c`?

Yes, sort of. The `[Makefile](https://github.com/penguian/access-esm-build-gadi/blob/master/Makefile )` uses the NCI `netcdf/4.7.1` module which includes both the equivalent of the `netcdf-c` and `netcdf-fortran` Spack packages. See the directory  `/apps/netcdf/4.7.1` for details.

**Comment by CodeGat** [comment]
@harshula in reference to:
> The Makefile uses the NCI netcdf/4.7.1 module which includes both the equivalent of the netcdf-c and netcdf-fortran Spack packages. See the directory /apps/netcdf/4.7.1 for details.

Would we be better off using the system `netcdf` rather than the spack-packages one?

**Comment by harshula** [comment]
> Would we be better off using the system `netcdf` rather than the spack-packages one?

No. We decided to avoid using the system `netcdf`.

**Comment by penguian** [comment]
Also:
```
$ spack install netcdf-fortran@4.5.1%intel@19.0.5.281 arch=linux-rocky8-x86_64
==> Error: concretization failed for the following reasons:

   1. Cannot satisfy 'netcdf-fortran@4.5.1'
   2. Cannot satisfy 'netcdf-fortran@4.5.1'
```

**Comment by CodeGat** [comment]
Noted, we are using `@4.6.1`

File: spack.yaml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+#
+# It describes a set of packages to be installed, along with
+# configuration settings.
+spack:
+  specs:
+    - access-esm1p5@git.2024.05.0
+  packages:
+    # Direct ACCESS-NRI dependencies
+    # Note: mom5, oasis3-mct have branch-specific logic and hence can't use
+    # the usual '@git.DATE' format.
```
Comment (aidanheerdegen) [inline spack.yaml:11]:
```suggestion
    # the usual '@git.DATE' ([calver](https://calver.org)) format.
```

Doesn't the `um` also use the branch specification?

**Comment by CodeGat** [comment]
Linking https://github.com/ACCESS-NRI/spack-packages/pull/121 for possible changes to the `spack.yaml`.

File: spack.yaml (hunk 1 above)
Comment (CodeGat) [reply spack.yaml]: Done in https://github.com/ACCESS-NRI/ACCESS-ESM1.5/pull/5/commits/34ed057cfc8f8dad6819b4194d1ec4caa5ce52e4

**Comment by harshula** [comment]
Hi @CodeGat , See below for the relationship between `netcdf-c` and `netcdf-fortran` versions:
```
$ cat /apps/netcdf/4.7.1/lib/Intel/pkgconfig/netcdf-fortran.pc 
prefix=/apps/netcdf/4.7.1
exec_prefix=${prefix}
libdir=/apps/netcdf/4.7.1/lib/Intel
includedir=/apps/netcdf/4.7.1/include/Intel
ccompiler=gcc
fcompiler=ifort

Name: netcdf-fortran
Description: NetCDF Client Library for Fortran
URL: http://www.unidata.ucar.edu/netcdf
Version: 4.5.1
Requires.private: netcdf > 4.1.1
Libs: -L${libdir} -lnetcdff
Libs.private: -L${libdir} -lnetcdff -lnetcdf
Cflags: -I${includedir}
```
```
$ cat /apps/netcdf/4.7.4/lib/Intel/pkgconfig/netcdf-fortran.pc 
prefix=/apps/netcdf/4.7.4
exec_prefix=${prefix}
libdir=/apps/netcdf/4.7.4/lib/Intel
includedir=/apps/netcdf/4.7.4/include/Intel
ccompiler=gcc
fcompiler=ifort

Name: netcdf-fortran
Description: NetCDF Client Library for Fortran
URL: http://www.unidata.ucar.edu/netcdf
Version: 4.5.2
Requires.private: netcdf > 4.1.1
Libs: -L${libdir} -lnetcdff
Libs.private: -L${libdir} -lnetcdff -lnetcdf
Cflags: -I${includedir}
```

**Comment by CodeGat** [comment]
@harshula, how is it that @penguian ran the model using `netcdf-c@4.7.1` and `netcdf-fortran@4.5.2` in this PR: https://github.com/ACCESS-NRI/spack-packages/pull/121 ?

**Review by penguian — APPROVED** [review]
The executables created by this config have been tested with the `pre-industrial` configuration and appear to reproduce the output of the `coe` executables.
See https://github.com/ACCESS-NRI/access-esm1.5-configs/pull/26

**Review by penguian — APPROVED** [review]
Approved.

**Comment by CodeGat** [comment]
The approval has come quicker than expected! 
Lets not merge this until we move to `0.22` and https://github.com/ACCESS-NRI/ACCESS-ESM1.5/pull/7 and https://github.com/ACCESS-NRI/build-cd/issues/102 are merged.

**Comment by CodeGat** [comment]
Although the above installed, we still have issues with `module use`. Testing out ways to fix it now: https://github.com/ACCESS-NRI/ACCESS-ESM1.5/issues/8#issuecomment-2207906452

**Comment by CodeGat** [comment]
Finally! Modulefiles work as expected - they were using old modulefiles for some reason, but we deleted the old modulefiles recreated the current ones.

**Review by aidanheerdegen — DISMISSED** [review]
LGTM

File: spack.yaml
[the remark this answers is not in the document]
Comment (CodeGat) [reply spack.yaml:1]: That's great news. I suppose now we wait to be able to release it...

**Review by penguian — APPROVED** [review]
Approved.

**Pull request merged**